In [114]:
import pandas as pd
import numpy as np
from sklearn.preprocessing import MinMaxScaler
from sklearn.metrics.pairwise import cosine_similarity, euclidean_distances, manhattan_distances

In [115]:
# df = pd.read_csv('1980.csv')
df = pd.read_csv('data/DE/data.csv')

In [116]:
df.head()

name         artists                       album  \
0   That's No Way To Get Along  Robert Wilkins  The Original Rolling Stone   
1              I Got A Problem  Albert Collins                   Frostbite   
2  The Highway Is Like A Woman  Albert Collins                   Frostbite   
3                Alabama Blues  Robert Wilkins  The Original Rolling Stone   
4               Get Away Blues  Robert Wilkins  The Original Rolling Stone   

  release_date release_date_precision  chart_power              spotify_id  \
0   1980-01-01                    day          NaN  5JKdsNzhJGpoaKBFZpOKQ3   
1   1980-02-05                    day          NaN  6e2PhVePvqe7w4VNBTslef   
2   1980-02-05                    day          NaN  4Qzcy6jU0yeck7IuCqGWSi   
3   1980-01-01                    day          NaN  42zUiYaKltE7jFWb57fXAW   
4   1980-01-01                    day          NaN  0hCrfp9WImVO6KWI30O1sM   

                                    uri  popularity  danceability  ...  \
0  spotify:track:5JKdsNzhJGpoaKBFZpOKQ3          22         0.427  ...   
1  spotify:track:6e2PhVePvqe7w4VNBTslef          15         0.760  ...   
2  spotify:track:4Qzcy6jU0yeck7IuCqGWSi          16         0.751  ...   
3  spotify:track:42zUiYaKltE7jFWb57fXAW          11         0.490  ...   
4  spotify:track:0hCrfp9WImVO6KWI30O1sM           4         0.629  ...   

   speechiness  acousticness  instrumentalness  liveness  valence    tempo  \
0       0.0407         0.980           0.00000    0.1920    0.381   97.078   
1       0.0573         0.328           0.03570    0.0447    0.849  102.183   
2       0.0361         0.419           0.11000    0.1050    0.771   98.545   
3       0.0716         0.972           0.00000    0.1590    0.905  205.718   
4       0.0431         0.968           0.00238    0.0941    0.322  119.752   

   duration_ms  time_signature          isrc    genres  
0       173827               4  USA370640304  acoustic  
1       274707               4  USARL8071903  acoustic  
2       302600               4  USARL8071904  acoustic  
3       157933               3  USA370640301  acoustic  
4       213107               4  USA370640300  acoustic  

[5 rows x 24 columns]

In [117]:
df['genres'].unique()

array(['acoustic', 'afrobeat', 'alt-rock', 'alternative', 'ambient',
       'anime', 'bluegrass', 'blues', 'brazil', 'british', 'cantopop',
       'chicago-house', 'children', 'chill', 'classical', 'club',
       'comedy', 'country', 'dance', 'dancehall', 'disco', 'dub',
       'dubstep', 'edm', 'electro', 'electronic', 'folk', 'forro',
       'french', 'funk', 'garage', 'german', 'gospel', 'goth', 'groove',
       'grunge', 'guitar', 'hard-rock', 'hardcore', 'hardstyle',
       'heavy-metal', 'hip-hop', 'honky-tonk', 'house', 'indian', 'indie',
       'indie-pop', 'industrial', 'j-dance', 'j-idol', 'j-pop', 'j-rock',
       'jazz', 'k-pop', 'latin', 'latino', 'malay', 'mandopop', 'metal',
       'minimal-techno', 'mpb', 'new-age', 'opera', 'pagode', 'party',
       'piano', 'pop', 'pop-film', 'power-pop', 'progressive-house',
       'psych-rock', 'punk', 'punk-rock', 'r-n-b', 'reggae', 'reggaeton',
       'rock', 'rock-n-roll', 'rockabilly', 'salsa', 'samba', 'sertanejo',
       'sing

Check for duplicate values

In [118]:
df.shape

(349308, 24)

In [119]:
df.drop_duplicates(inplace=True)

In [120]:
df.shape

(83286, 24)

There were a lot of duplicate values..

Check for null values

In [121]:
df.isna().any()

name                      False
artists                   False
album                     False
release_date              False
release_date_precision    False
chart_power                True
spotify_id                False
uri                       False
popularity                False
danceability              False
energy                    False
key                       False
loudness                  False
mode                      False
speechiness               False
acousticness              False
instrumentalness          False
liveness                  False
valence                   False
tempo                     False
duration_ms               False
time_signature            False
isrc                      False
genres                    False
dtype: bool

There exist Null values for the Chart Power and for the isrc. The null values for the Chart Power are valid, because not every song was in the Charts. Lets have a look at the null values for isrc:

In [122]:
df[df['isrc'].isna()]

Empty DataFrame
Columns: [name, artists, album, release_date, release_date_precision, chart_power, spotify_id, uri, popularity, danceability, energy, key, loudness, mode, speechiness, acousticness, instrumentalness, liveness, valence, tempo, duration_ms, time_signature, isrc, genres]
Index: []

[0 rows x 24 columns]

Check for duplicate isrc numbers

In [123]:
df.columns

Index(['name', 'artists', 'album', 'release_date', 'release_date_precision',
       'chart_power', 'spotify_id', 'uri', 'popularity', 'danceability',
       'energy', 'key', 'loudness', 'mode', 'speechiness', 'acousticness',
       'instrumentalness', 'liveness', 'valence', 'tempo', 'duration_ms',
       'time_signature', 'isrc', 'genres'],
      dtype='object')

First group the songs by the spotify id. This brings together all instances with different genres but same spotify id.

In [391]:
by_spotify_id = df.copy()
by_spotify_id = by_spotify_id.groupby('spotify_id').agg({
    'genres': list,
    'name': list,
    'artists': list,
    'album': list,
    'release_date': list,
    'release_date_precision': list,
    'uri': list,
    'isrc':list,
    'chart_power': 'mean', 
    'popularity': 'mean', 
    'danceability': 'mean', 
    'energy': 'mean',
    'key': 'mean', 
    'loudness': 'mean', 
    'mode': 'mean', 
    'speechiness': 'mean', 
    'acousticness': 'mean', 
    'instrumentalness': 'mean', 
    'liveness': 'mean', 
    'valence': 'mean',
    'tempo': 'mean', 
    'duration_ms': 'mean', 
    'time_signature': 'mean'
})

In [392]:
by_spotify_id.head()

genres  \
spotify_id                                    
000P83HDtOHcNVFZy7Q2Yu              [salsa]   
000paUq6xoYcvCz2REJdu1      [pop, pop-film]   
0017A9myIXhXYB4gBsPRsA                [pop]   
001reycpvPkUjmpxTJv6eV  [classical, german]   
002H0J6N0wKjjRtFQhKDzZ     [dance, turkish]   

                                                                     name  \
spotify_id                                                                  
000P83HDtOHcNVFZy7Q2Yu                           [El Licor de Tu Boquita]   
000paUq6xoYcvCz2REJdu1  [Hun Tey Kharkan Gay Alarum, Hun Tey Kharkan G...   
0017A9myIXhXYB4gBsPRsA                                [Untuk Sebuah Nama]   
001reycpvPkUjmpxTJv6eV  [Oboe Concerto in F Minor, TWV 51:f1: I. Alleg...   
002H0J6N0wKjjRtFQhKDzZ         [Arabesk Oyun Havası, Arabesk Oyun Havası]   

                                                                  artists  \
spotify_id                                                                  
000P83HDtOHcNVFZy7Q2Yu                     [El Gran Combo De Puerto Rico]   
000paUq6xoYcvCz2REJdu1                           [Noor Jehan, Noor Jehan]   
0017A9myIXhXYB4gBsPRsA                                   [Ratih Purwasih]   
001reycpvPkUjmpxTJv6eV  [Georg Philipp Telemann,Burkhard Glaetzner,Har...   
002H0J6N0wKjjRtFQhKDzZ     [Esin Engin Orkestrası, Esin Engin Orkestrası]   

                                                                    album  \
spotify_id                                                                  
000P83HDtOHcNVFZy7Q2Yu                                            [Unity]   
000paUq6xoYcvCz2REJdu1  [Noor Jehan New Punjabi Songs, Noor Jehan New ...   
0017A9myIXhXYB4gBsPRsA                                [Untuk Sebuah Nama]   
001reycpvPkUjmpxTJv6eV  [Oboe Concertos (Baroque) - Antonio Vivaldi: G...   
002H0J6N0wKjjRtFQhKDzZ                           [Arabesk'80, Arabesk'80]   

                                    release_date release_date_precision  \
spotify_id                                                                
000P83HDtOHcNVFZy7Q2Yu                    [1980]                 [year]   
000paUq6xoYcvCz2REJdu1  [1980-12-01, 1980-12-01]             [day, day]   
0017A9myIXhXYB4gBsPRsA                    [1980]                 [year]   
001reycpvPkUjmpxTJv6eV              [1980, 1980]           [year, year]   
002H0J6N0wKjjRtFQhKDzZ              [1980, 1980]           [year, year]   

                                                                      uri  \
spotify_id                                                                  
000P83HDtOHcNVFZy7Q2Yu             [spotify:track:000P83HDtOHcNVFZy7Q2Yu]   
000paUq6xoYcvCz2REJdu1  [spotify:track:000paUq6xoYcvCz2REJdu1, spotify...   
0017A9myIXhXYB4gBsPRsA             [spotify:track:0017A9myIXhXYB4gBsPRsA]   
001reycpvPkUjmpxTJv6eV  [spotify:track:001reycpvPkUjmpxTJv6eV, spotify...   
002H0J6N0wKjjRtFQhKDzZ  [spotify:track:002H0J6N0wKjjRtFQhKDzZ, spotify...   

                                                isrc  chart_power  popularity  \
spotify_id                                                                      
000P83HDtOHcNVFZy7Q2Yu                [USMRE0801806]          NaN        21.0   
000paUq6xoYcvCz2REJdu1  [USA561127553, USA561127553]          NaN         0.0   
0017A9myIXhXYB4gBsPRsA                [UKFT41826340]          NaN         0.0   
001reycpvPkUjmpxTJv6eV  [DES380744015, DES380744015]          NaN         0.0   
002H0J6N0wKjjRtFQhKDzZ  [BGA472127386, BGA472127386]          NaN         0.0   

                        ...  loudness  mode  speechiness  acousticness  \
spotify_id              ...                                              
000P83HDtOHcNVFZy7Q2Yu  ...    -8.377   0.0       0.0736         0.438   
000paUq6xoYcvCz2REJdu1  ...    -9.944   1.0       0.0868         0.741   
0017A9myIXhXYB4gBsPRsA  ...    -3.549   1.0       0.0289         0.084   
001reycpvPkUjmpxTJv6eV  ...   -14.759   0.0       0.0348         0.831   
002H0J6N0

Then reduce all categorical features such that there are no duplicated values in an instance for one feature.

In [366]:
def reduce_list(elements, string_return = True):
    '''
    Removes duplicate elements in a list

    Parameter
    ---------
    elements: list
        List that should be reduced

    string_return: boolean; default=True
        Whether a list with just one element should be returned as string or list

    Return
    ------
    unique_elements: list or str
    
    '''
    unique_elements = []
    for element in elements:
        if element not in unique_elements:
            unique_elements.append(element)
    if (len(unique_elements) == 1 and string_return):
        return unique_elements[0]
    return unique_elements

In [367]:
by_spotify_id['isrc'] = by_spotify_id['isrc'].agg(reduce_list)
by_spotify_id['name'] = by_spotify_id['name'].agg(reduce_list)
by_spotify_id['artists'] = by_spotify_id['artists'].agg(reduce_list)
by_spotify_id['album'] = by_spotify_id['album'].agg(reduce_list)
by_spotify_id['release_date'] = by_spotify_id['release_date'].agg(reduce_list)
by_spotify_id['release_date_precision'] = by_spotify_id['release_date_precision'].agg(reduce_list)
by_spotify_id['uri'] = by_spotify_id['uri'].agg(reduce_list)
by_spotify_id['genres'] = by_spotify_id['genres'].agg(reduce_list)

In [368]:
by_spotify_id.head()

genres  \
spotify_id                                    
000P83HDtOHcNVFZy7Q2Yu                salsa   
000paUq6xoYcvCz2REJdu1      [pop, pop-film]   
0017A9myIXhXYB4gBsPRsA                  pop   
001reycpvPkUjmpxTJv6eV  [classical, german]   
002H0J6N0wKjjRtFQhKDzZ     [dance, turkish]   

                                                                   name  \
spotify_id                                                                
000P83HDtOHcNVFZy7Q2Yu                           El Licor de Tu Boquita   
000paUq6xoYcvCz2REJdu1                       Hun Tey Kharkan Gay Alarum   
0017A9myIXhXYB4gBsPRsA                                Untuk Sebuah Nama   
001reycpvPkUjmpxTJv6eV  Oboe Concerto in F Minor, TWV 51:f1: I. Allegro   
002H0J6N0wKjjRtFQhKDzZ                              Arabesk Oyun Havası   

                                                                  artists  \
spotify_id                                                                  
000P83HDtOHcNVFZy7Q2Yu                       El Gran Combo De Puerto Rico   
000paUq6xoYcvCz2REJdu1                                         Noor Jehan   
0017A9myIXhXYB4gBsPRsA                                     Ratih Purwasih   
001reycpvPkUjmpxTJv6eV  Georg Philipp Telemann,Burkhard Glaetzner,Hart...   
002H0J6N0wKjjRtFQhKDzZ                              Esin Engin Orkestrası   

                                                                    album  \
spotify_id                                                                  
000P83HDtOHcNVFZy7Q2Yu                                              Unity   
000paUq6xoYcvCz2REJdu1                       Noor Jehan New Punjabi Songs   
0017A9myIXhXYB4gBsPRsA                                  Untuk Sebuah Nama   
001reycpvPkUjmpxTJv6eV  Oboe Concertos (Baroque) - Antonio Vivaldi: Ge...   
002H0J6N0wKjjRtFQhKDzZ                                         Arabesk'80   

                       release_date release_date_precision  \
spotify_id                                                   
000P83HDtOHcNVFZy7Q2Yu         1980                   year   
000paUq6xoYcvCz2REJdu1   1980-12-01                    day   
0017A9myIXhXYB4gBsPRsA         1980                   year   
001reycpvPkUjmpxTJv6eV         1980                   year   
002H0J6N0wKjjRtFQhKDzZ         1980                   year   

                                                         uri          isrc  \
spotify_id                                                                   
000P83HDtOHcNVFZy7Q2Yu  spotify:track:000P83HDtOHcNVFZy7Q2Yu  USMRE0801806   
000paUq6xoYcvCz2REJdu1  spotify:track:000paUq6xoYcvCz2REJdu1  USA561127553   
0017A9myIXhXYB4gBsPRsA  spotify:track:0017A9myIXhXYB4gBsPRsA  UKFT41826340   
001reycpvPkUjmpxTJv6eV  spotify:track:001reycpvPkUjmpxTJv6eV  DES380744015   
002H0J6N0wKjjRtFQhKDzZ  spotify:track:002H0J6N0wKjjRtFQhKDzZ  BGA472127386   

                        chart_power  popularity  ...  loudness  mode  \
spotify_id                                       ...                   
000P83HDtOHcNVFZy7Q2Yu          NaN        21.0  ...    -8.377   0.0   
000paUq6xoYcvCz2REJdu1          NaN         0.0  ...    -9.944   1.0   
0017A9myIXhXYB4gBsPRsA          NaN         0.0  ...    -3.549   1.0   
001reycpvPkUjmpxTJv6eV          NaN         0.0  ...   -14.759   0.0   
002H0J6N0wKjjRtFQhKDzZ          NaN         0.0  ...   -12.444   0.0   

                        speechiness  acousticness  instrumentalness  liveness  \
spotify_id                                                                      
000P83HDtOHcNVFZy7Q2Yu       0.0736         0.438          0.000042     0.126   
000paUq6xoYcvCz2REJdu1       0.0868         0.741          0.037700     0.222   
0017A9myIXhXYB4gBsPRsA       0.0289         0.084          0.000004     0.117   
001reycpvPkUjmpxTJv6eV       0.0348         0.831          0.916000     0.129   
002H0J6N0wKjjRtFQhKDzZ       0.0365         0.528          0.323000     0.343   

                        valence    tempo  dura

The next step is to group all the songs by the isrc number since this should be the unique identifier. 

In [378]:
by_isrc = by_spotify_id.copy()
by_isrc = by_isrc[['isrc', 'name']]
by_isrc['count'] = 0
by_isrc = by_isrc.groupby('isrc').agg({
    'name': list,
    'count': 'count'
})

by_isrc.sort_values('count', ascending=False)

name  count
isrc                                                                  
USSM18000294  [Mallorca (Barcarola), Op. 202 [Arranged by Jo...      7
USSM18000291  [Suite Española No. 1, Op. 47: No. 1, Granada ...      7
USSM18000295  [Cantos de España, Op. 232: No. 4, Córdoba (Ar...      7
USSM18000293  [Suite Española No. 1, Op. 47: No. 3, Sevilla ...      6
USSM18000292  [Suite Española No. 1, Op. 47: No. 5, Asturias...      6
...                                                         ...    ...
GB2DY1000060                                  [Baphomet - Demo]      1
GB2DY1000061                                 [Sorcerers - Demo]      1
GB2DY1000062                         [Extermination Day - Demo]      1
GB2DY1000063                                 [Flight 19 - Demo]      1
usx9p2035280                               [Tip of the Iceberg]      1

[45608 rows x 2 columns]

In [379]:
by_isrc['name'] = by_isrc['name'].agg(reduce_list, string_return=False)

In [380]:
special_cases = by_isrc[by_isrc['name'].apply(len) > 1]

In [383]:
special_cases['name'] = special_cases['name'].apply(reduce_list)

/var/folders/t4/w1slb7fj5kg0tc41jhrm3zx00000gn/T/ipykernel_2586/102079820.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  special_cases['name'] = special_cases['name'].apply(reduce_list)


In [384]:
special_cases

name  count
isrc                                                                  
BRGVR1900214              [Comparação, Conversão do Padre João]      2
BRGVR1900215                 [O Bom Cristão, Eu Vou para o Céu]      2
BRGVR1900216                          [Eu Sou, Jesus de Nazaré]      2
BRGVR1900217                     [Escondido em Jesus, Vencedor]      2
BRGVR1900218               [As Duas Ovelhas, Alma Transformada]      2
...                                                         ...    ...
USSM10411835  [Cello Concerto No. 2 in D Major, Hob. VIIb:2:...      5
USSM10411836  [Cello Concerto No. 2 in D Major, Hob. VIIb:2:...      5
USSM17900406  [Cello Concerto No. 1 in C Major, Hob. VIIb:1:...      4
USSM17900407  [Concerto in C Major for Cello and Orchestra: ...      4
USSM17900408  [Cello Concerto No. 1 in C Major, Hob. VIIb:1:...      4

[311 rows x 2 columns]

In [385]:
indices = special_cases.index


In [389]:
by_isrc.drop(index=indices.values)

name  count
isrc                                        
ARA821922001           [Tierra Negra]      1
ARA821922002            [El Incendio]      1
ARA821922003   [Patotero Sentimental]      1
ARA821922004            [A Media Luz]      1
ARA821922005    [Portero Suba y Diga]      1
...                               ...    ...
usx9p2035276  [Bad News Travels Fast]      1
usx9p2035277            [All Tied Up]      1
usx9p2035278          [Givin' It All]      1
usx9p2035279    [It May Never Happen]      1
usx9p2035280     [Tip of the Iceberg]      1

[45297 rows x 2 columns]

Group the DataFrame by the spotify id. All catgeorical values of the features are stored as lists. All numerical features are stored as the mean number.

In [100]:
df_total = df.copy()

df_total = df_total.groupby('spotify_id').agg({
    'spotify_id': list,
    'genres': list,
    'name': list,
    'artists': list,
    'album': list,
    'release_date': list,
    'release_date_precision': list,
    'uri': list,
    'isrc':list,
    'chart_power': 'mean', 
    'popularity': 'mean', 
    'danceability': 'mean', 
    'energy': 'mean',
    'key': 'mean', 
    'loudness': 'mean', 
    'mode': 'mean', 
    'speechiness': 'mean', 
    'acousticness': 'mean', 
    'instrumentalness': 'mean', 
    'liveness': 'mean', 
    'valence': 'mean',
    'tempo': 'mean', 
    'duration_ms': 'mean', 
    'time_signature': 'mean'
})

In [106]:
display(df_total.head())
df_total.shape

spotify_id  \
spotify_id                                                                  
000P83HDtOHcNVFZy7Q2Yu                           [000P83HDtOHcNVFZy7Q2Yu]   
000blRZoW279EJlePCHsoi   [000blRZoW279EJlePCHsoi, 000blRZoW279EJlePCHsoi]   
000paUq6xoYcvCz2REJdu1   [000paUq6xoYcvCz2REJdu1, 000paUq6xoYcvCz2REJdu1]   
0017A9myIXhXYB4gBsPRsA                           [0017A9myIXhXYB4gBsPRsA]   
0017lMteokXc5Mzcr2BYFp  [0017lMteokXc5Mzcr2BYFp, 0017lMteokXc5Mzcr2BYF...   

                                            genres  \
spotify_id                                           
000P83HDtOHcNVFZy7Q2Yu                     [salsa]   
000blRZoW279EJlePCHsoi                 [folk, pop]   
000paUq6xoYcvCz2REJdu1             [pop, pop-film]   
0017A9myIXhXYB4gBsPRsA                       [pop]   
0017lMteokXc5Mzcr2BYFp  [classical, french, opera]   

                                                                     name  \
spotify_id                                                                  
000P83HDtOHcNVFZy7Q2Yu                           [El Licor de Tu Boquita]   
000blRZoW279EJlePCHsoi  [La hermosura de una flor, La hermosura de una...   
000paUq6xoYcvCz2REJdu1  [Hun Tey Kharkan Gay Alarum, Hun Tey Kharkan G...   
0017A9myIXhXYB4gBsPRsA                                [Untuk Sebuah Nama]   
0017lMteokXc5Mzcr2BYFp  [Werther / Act 1: "Assez! Assez!" - "Noël! Jés...   

                                                                  artists  \
spotify_id                                                                  
000P83HDtOHcNVFZy7Q2Yu                     [El Gran Combo De Puerto Rico]   
000blRZoW279EJlePCHsoi                 [Jose Ali Nieves, Jose Ali Nieves]   
000paUq6xoYcvCz2REJdu1                           [Noor Jehan, Noor Jehan]   
0017A9myIXhXYB4gBsPRsA                                   [Ratih Purwasih]   
0017lMteokXc5Mzcr2BYFp  [Jules Massenet,Robert Lloyd,Malcolm King,Paul...   

                                                                    album  \
spotify_id                                                                  
000P83HDtOHcNVFZy7Q2Yu                                            [Unity]   
000blRZoW279EJlePCHsoi  [Estampas De Puerto Ayacucho, Estampas De Puer...   
000paUq6xoYcvCz2REJdu1  [Noor Jehan New Punjabi Songs, Noor Jehan New ...   
0017A9myIXhXYB4gBsPRsA                                [Untuk Sebuah Nama]   
0017lMteokXc5Mzcr2BYFp  [Massenet: Werther, Massenet: Werther, Massene...   

                                                release_date  \
spotify_id                                                     
000P83HDtOHcNVFZy7Q2Yu                                [1980]   
000blRZoW279EJlePCHsoi              [1981-04-30, 1981-04-30]   
000paUq6xoYcvCz2REJdu1              [1980-12-01, 1980-12-01]   
0017A9myIXhXYB4gBsPRsA                                [1980]   
0017lMteokXc5Mzcr2BYFp  [1981-01-01, 1981-01-01, 1981-01-01]   

                       release_date_precision  \
spotify_id                                      
000P83HDtOHcNVFZy7Q2Yu                 [year]   
000blRZoW279EJlePCHsoi             [day, day]   
000paUq6xoYcvCz2REJdu1             [day, day]   
0017A9myIXhXYB4gBsPRsA                 [year]   
0017lMteokXc5Mzcr2BYFp        [day, day, day]   

                                                                      uri  \
spotify_id                                                                  
000P83HDtOHcNVFZy7Q2Yu             [spotify:track:000P83HDtOHcNVFZy7Q2Yu]   
000blRZoW279EJlePCHsoi  [spotify:track:000blRZoW279EJlePCHsoi, spotify...   
000paUq6xoYcvCz2REJdu1  [spotify:track:000paUq6xoYcvCz2REJdu1, spotify...   
0017A9myIXhXYB4gBsPRsA             [spotify:track:0017A9myIXhXYB4gBsPRsA]   
0017lMteokXc5Mzcr2BYFp  [spotify:track:0017lMteokXc5Mzcr2BYFp, spotify...   

                                                              isrc  \
spotify_id                                                           
000P83HDtOHcNVFZy7Q2Yu                             

(77436, 24)

Create a DataFrame which contains only the relevant features for the recommender system

In [22]:
df_recommender = df.drop(columns=['artists', 'genres','album', 'release_date', 'release_date_precision', 'chart_power', 'uri', 'popularity', 'name', 'isrc'])

Drop duplicate values..

In [23]:
df_recommender.shape

(76414, 14)

In [24]:
df_recommender.drop_duplicates(inplace=True)

In [25]:
df_recommender.shape

(38456, 14)

There were a lot of duplicates..

In [26]:
df_recommender.head()

spotify_id  danceability  energy  key  loudness  mode  \
0  5xgPirJuHEYxM3pMFPYtXG         0.600   0.118    3   -18.452     1   
1  6vonkFQZRt9WHHfO6iU1AL         0.887   0.344    1   -11.985     1   
2  7pFvU6cr3CURlNc8S9SxAp         0.433   0.197    8   -13.150     1   
3  6CR1NaaMBPQUed2i5QRW4l         0.442   0.167    2   -14.996     1   
4  1td7cg3olnhudXRKSASjRZ         0.468   0.205    2   -13.606     0   

   speechiness  acousticness  instrumentalness  liveness  valence    tempo  \
0       0.0402         0.977          0.072900    0.1100    0.557   98.429   
1       0.0481         0.832          0.028000    0.3370    0.620  129.867   
2       0.0324         0.993          0.003270    0.1090    0.804   85.316   
3       0.0358         0.859          0.001730    0.0876    0.243   72.212   
4       0.0491         0.778          0.000237    0.0832    0.374  114.904   

   duration_ms  time_signature  
0       278787               4  
1       171445               4  
2       155840               4  
3       189427               4  
4       209107               3

Since all features except the spotify_id are numerical features and the spotify_id should be an unique identifier, try to group by the id to reduce the DataFrame even more.

In [27]:
df_recommender = df_recommender.groupby(['spotify_id']).mean()

In [28]:
display(df_recommender.head())
df_recommender.shape

danceability  energy  key  loudness  mode  \
spotify_id                                                          
000P83HDtOHcNVFZy7Q2Yu         0.619   0.561  7.0    -8.377   0.0   
000paUq6xoYcvCz2REJdu1         0.492   0.558  2.0    -9.944   1.0   
0017A9myIXhXYB4gBsPRsA         0.220   0.668  0.0    -3.549   1.0   
001reycpvPkUjmpxTJv6eV         0.448   0.224  5.0   -14.759   0.0   
002H0J6N0wKjjRtFQhKDzZ         0.286   0.480  2.0   -12.444   0.0   

                        speechiness  acousticness  instrumentalness  liveness  \
spotify_id                                                                      
000P83HDtOHcNVFZy7Q2Yu       0.0736         0.438          0.000042     0.126   
000paUq6xoYcvCz2REJdu1       0.0868         0.741          0.037700     0.222   
0017A9myIXhXYB4gBsPRsA       0.0289         0.084          0.000004     0.117   
001reycpvPkUjmpxTJv6eV       0.0348         0.831          0.916000     0.129   
002H0J6N0wKjjRtFQhKDzZ       0.0365         0.528          0.323000     0.343   

                        valence    tempo  duration_ms  time_signature  
spotify_id                                                             
000P83HDtOHcNVFZy7Q2Yu    0.696   88.940     242493.0             4.0  
000paUq6xoYcvCz2REJdu1    0.606   96.169     372167.0             4.0  
0017A9myIXhXYB4gBsPRsA    0.360   81.433     249504.0             4.0  
001reycpvPkUjmpxTJv6eV    0.540  112.185     215733.0             4.0  
002H0J6N0wKjjRtFQhKDzZ    0.752  134.274     345943.0             4.0

(38456, 13)

The shape stayed the same, so there were no duplicate ids.

Scale the data, so every feature has the same influence.

In [29]:
scaler = MinMaxScaler()
df_recommender_scaled = scaler.fit_transform(df_recommender)
df_recommender_scaled = pd.DataFrame(df_recommender_scaled, columns=df_recommender.columns, index = df_recommender.index)

In [30]:
df_recommender_scaled.head()

danceability    energy       key  loudness  mode  \
spotify_id                                                                 
000P83HDtOHcNVFZy7Q2Yu      0.638803  0.561519  0.636364  0.798337   0.0   
000paUq6xoYcvCz2REJdu1      0.507740  0.558516  0.181818  0.764918   1.0   
0017A9myIXhXYB4gBsPRsA      0.227038  0.668637  0.000000  0.901301   1.0   
001reycpvPkUjmpxTJv6eV      0.462332  0.224149  0.454545  0.662231   0.0   
002H0J6N0wKjjRtFQhKDzZ      0.295150  0.480430  0.181818  0.711602   0.0   

                        speechiness  acousticness  instrumentalness  liveness  \
spotify_id                                                                      
000P83HDtOHcNVFZy7Q2Yu     0.076507      0.439758          0.000042  0.114847   
000paUq6xoYcvCz2REJdu1     0.090229      0.743976          0.037851  0.212072   
0017A9myIXhXYB4gBsPRsA     0.030042      0.084336          0.000004  0.105732   
001reycpvPkUjmpxTJv6eV     0.036175      0.834337          0.919679  0.117885   
002H0J6N0wKjjRtFQhKDzZ     0.037942      0.530120          0.324297  0.334616   

                        valence     tempo  duration_ms  time_signature  
spotify_id                                                              
000P83HDtOHcNVFZy7Q2Yu    0.696  0.387169     0.067168             0.8  
000paUq6xoYcvCz2REJdu1    0.606  0.418638     0.104717             0.8  
0017A9myIXhXYB4gBsPRsA    0.360  0.354490     0.069198             0.8  
001reycpvPkUjmpxTJv6eV    0.540  0.488358     0.059419             0.8  
002H0J6N0wKjjRtFQhKDzZ    0.752  0.584514     0.097124             0.8

In [31]:
df_recommender_scaled.reset_index(inplace=True)
df_recommender.reset_index(inplace=True)

Convert every genre to a feature. If a song is part of a genre it should contain the value 1 otherwise 0.

In [32]:
ct = pd.crosstab(df['spotify_id'], df['genres'])
ct.reset_index(inplace=True)

In [33]:
display(ct.head())
ct.shape

genres              spotify_id  acoustic  afrobeat  alt-rock  alternative  \
0       000P83HDtOHcNVFZy7Q2Yu         0         0         0            0   
1       000paUq6xoYcvCz2REJdu1         0         0         0            0   
2       0017A9myIXhXYB4gBsPRsA         0         0         0            0   
3       001reycpvPkUjmpxTJv6eV         0         0         0            0   
4       002H0J6N0wKjjRtFQhKDzZ         0         0         0            0   

genres  ambient  anime  bluegrass  blues  brazil  ...  sleep  songwriter  \
0             0      0          0      0       0  ...      0           0   
1             0      0          0      0       0  ...      0           0   
2             0      0          0      0       0  ...      0           0   
3             0      0          0      0       0  ...      0           0   
4             0      0          0      0       0  ...      0           0   

genres  soul  spanish  swedish  synth-pop  tango  techno  trance  turkish  
0          0        0        0          0      0       0       0        0  
1          0        0        0          0      0       0       0        0  
2          0        0        0          0      0       0       0        0  
3          0        0        0          0      0       0       0        0  
4          0        0        0          0      0       0       0        1  

[5 rows x 96 columns]

(38456, 96)

Merge both DataFrames together to create the Recommender System.

In [34]:
ct_merged = ct.merge(df_recommender_scaled, on=['spotify_id'], how='left')

In [35]:
ct_merged.set_index(['spotify_id'], inplace=True)

In [36]:
display(ct_merged.head())
ct_merged.shape

acoustic  afrobeat  alt-rock  alternative  ambient  \
spotify_id                                                                   
000P83HDtOHcNVFZy7Q2Yu         0         0         0            0        0   
000paUq6xoYcvCz2REJdu1         0         0         0            0        0   
0017A9myIXhXYB4gBsPRsA         0         0         0            0        0   
001reycpvPkUjmpxTJv6eV         0         0         0            0        0   
002H0J6N0wKjjRtFQhKDzZ         0         0         0            0        0   

                        anime  bluegrass  blues  brazil  british  ...  \
spotify_id                                                        ...   
000P83HDtOHcNVFZy7Q2Yu      0          0      0       0        0  ...   
000paUq6xoYcvCz2REJdu1      0          0      0       0        0  ...   
0017A9myIXhXYB4gBsPRsA      0          0      0       0        0  ...   
001reycpvPkUjmpxTJv6eV      0          0      0       0        0  ...   
002H0J6N0wKjjRtFQhKDzZ      0          0      0       0        0  ...   

                        loudness  mode  speechiness  acousticness  \
spotify_id                                                          
000P83HDtOHcNVFZy7Q2Yu  0.798337   0.0     0.076507      0.439758   
000paUq6xoYcvCz2REJdu1  0.764918   1.0     0.090229      0.743976   
0017A9myIXhXYB4gBsPRsA  0.901301   1.0     0.030042      0.084336   
001reycpvPkUjmpxTJv6eV  0.662231   0.0     0.036175      0.834337   
002H0J6N0wKjjRtFQhKDzZ  0.711602   0.0     0.037942      0.530120   

                        instrumentalness  liveness  valence     tempo  \
spotify_id                                                              
000P83HDtOHcNVFZy7Q2Yu          0.000042  0.114847    0.696  0.387169   
000paUq6xoYcvCz2REJdu1          0.037851  0.212072    0.606  0.418638   
0017A9myIXhXYB4gBsPRsA          0.000004  0.105732    0.360  0.354490   
001reycpvPkUjmpxTJv6eV          0.919679  0.117885    0.540  0.488358   
002H0J6N0wKjjRtFQhKDzZ          0.324297  0.334616    0.752  0.584514   

                        duration_ms  time_signature  
spotify_id                                           
000P83HDtOHcNVFZy7Q2Yu     0.067168             0.8  
000paUq6xoYcvCz2REJdu1     0.104717             0.8  
0017A9myIXhXYB4gBsPRsA     0.069198             0.8  
001reycpvPkUjmpxTJv6eV     0.059419             0.8  
002H0J6N0wKjjRtFQhKDzZ     0.097124             0.8  

[5 rows x 108 columns]

(38456, 108)

Overall there are 108 features used for the Recommendation system.

Try different distance measures / similarity functions.

In [37]:
from sklearn.metrics.pairwise import cosine_similarity, linear_kernel, polynomial_kernel, sigmoid_kernel

Compute the cosine similarity between each song. A value close to one means that 2 two songs are equal a value close to 0 means, that they are completley different.

Cosine similarity

In [38]:
cosine_similarity_array = cosine_similarity(ct_merged)

In [39]:
cosine_similarity_array.shape

(38456, 38456)

In [40]:
cosine_similarity_df = pd.DataFrame(cosine_similarity_array, index=ct_merged.index, columns=ct_merged.index)

In [41]:
cosine_similarity_df

spotify_id              000P83HDtOHcNVFZy7Q2Yu  000paUq6xoYcvCz2REJdu1  \
spotify_id                                                               
000P83HDtOHcNVFZy7Q2Yu                1.000000                0.584950   
000paUq6xoYcvCz2REJdu1                0.584950                1.000000   
0017A9myIXhXYB4gBsPRsA                0.547657                0.850264   
001reycpvPkUjmpxTJv6eV                0.577832                0.483625   
002H0J6N0wKjjRtFQhKDzZ                0.615026                0.521155   
...                                        ...                     ...   
7zsH8mPq4nPBqqqDJpjgJe                0.656446                0.723518   
7zv3sIP0Umtz9MswoJSF1T                0.619765                0.538335   
7zwHfIlTAIPa6hgjzK1MGx                0.633239                0.676929   
7zwlOlX3MA7BFkmf0c1y4p                0.511864                0.567409   
7zzdiBe3usdBtyqHEekCEs                0.490575                0.531816   

spotify_id              0017A9myIXhXYB4gBsPRsA  001reycpvPkUjmpxTJv6eV  \
spotify_id                                                               
000P83HDtOHcNVFZy7Q2Yu                0.547657                0.577832   
000paUq6xoYcvCz2REJdu1                0.850264                0.483625   
0017A9myIXhXYB4gBsPRsA                1.000000                0.398664   
001reycpvPkUjmpxTJv6eV                0.398664                1.000000   
002H0J6N0wKjjRtFQhKDzZ                0.490335                0.554062   
...                                        ...                     ...   
7zsH8mPq4nPBqqqDJpjgJe                0.790129                0.491403   
7zv3sIP0Umtz9MswoJSF1T                0.437428                0.569164   
7zwHfIlTAIPa6hgjzK1MGx                0.627054                0.656092   
7zwlOlX3MA7BFkmf0c1y4p                0.495387                0.570760   
7zzdiBe3usdBtyqHEekCEs                0.591632                0.369105   

spotify_id              002H0J6N0wKjjRtFQhKDzZ  002Ra2MFY7R5ci6q1YLWLH  \
spotify_id                                                               
000P83HDtOHcNVFZy7Q2Yu                0.615026                0.733066   
000paUq6xoYcvCz2REJdu1                0.521155                0.757498   
0017A9myIXhXYB4gBsPRsA                0.490335                0.702231   
001reycpvPkUjmpxTJv6eV                0.554062                0.586211   
002H0J6N0wKjjRtFQhKDzZ                1.000000                0.576255   
...                                        ...                     ...   
7zsH8mPq4nPBqqqDJpjgJe                0.551722                0.853721   
7zv3sIP0Umtz9MswoJSF1T                0.543438                0.659313   
7zwHfIlTAIPa6hgjzK1MGx                0.599098                0.614871   
7zwlOlX3MA7BFkmf0c1y4p                0.459383                0.558395   
7zzdiBe3usdBtyqHEekCEs                0.575584                0.644105   

spotify_id              003787JNriexFRPKRwyMTn  003FCWl5ci9MxCaiTbErSr  \
spotify_id                                                               
000P83HDtOHcNVFZy7Q2Yu                0.436763                0.696065   
000paUq6xoYcvCz2REJdu1                0.552569                0.725537   
0017A9myIXhXYB4gBsPRsA                0.499048                0.689698   
001reycpvPkUjmpxTJv6eV                0.539577                0.549387   
002H0J6N0wKjjRtFQhKDzZ                0.427358                0.589067   
...                                        ...                     ...   
7zsH8mPq4nPBqqqDJpjgJe                0.366343                0.582482   
7zv3sIP0Umtz9MswoJSF1T                0.482791                0.584925   
7zwHfIlTAIPa6hgjzK1MGx                0.688482                0.765184   
7zwlOlX3MA7BFkmf0c1y4p                0.814217                0.625960   
7zzdiBe3usdBtyqHEekCEs                0.286651                0.419448   

spotify_id              004223m8Zb9ZHcxHFxlpQc  006XjVYo0HbGrUfvL4FodP  ...  \
spotify_id                                                           

Linear kernel

In [42]:
linear_kernel_array = linear_kernel(ct_merged)

In [43]:
linear_kernel_array.shape

(38456, 38456)

In [44]:
linear_kernel_df = pd.DataFrame(linear_kernel_array, index=ct_merged.index, columns=ct_merged.index)
linear_kernel_df.head()

spotify_id              000P83HDtOHcNVFZy7Q2Yu  000paUq6xoYcvCz2REJdu1  \
spotify_id                                                               
000P83HDtOHcNVFZy7Q2Yu                4.256930                2.953650   
000paUq6xoYcvCz2REJdu1                2.953650                5.989404   
0017A9myIXhXYB4gBsPRsA                2.324010                4.279828   
001reycpvPkUjmpxTJv6eV                2.831302                2.810841   
002H0J6N0wKjjRtFQhKDzZ                2.812805                2.827201   

spotify_id              0017A9myIXhXYB4gBsPRsA  001reycpvPkUjmpxTJv6eV  \
spotify_id                                                               
000P83HDtOHcNVFZy7Q2Yu                2.324010                2.831302   
000paUq6xoYcvCz2REJdu1                4.279828                2.810841   
0017A9myIXhXYB4gBsPRsA                4.230210                1.947260   
001reycpvPkUjmpxTJv6eV                1.947260                5.639912   
002H0J6N0wKjjRtFQhKDzZ                2.235485                2.916704   

spotify_id              002H0J6N0wKjjRtFQhKDzZ  002Ra2MFY7R5ci6q1YLWLH  \
spotify_id                                                               
000P83HDtOHcNVFZy7Q2Yu                2.812805                3.180754   
000paUq6xoYcvCz2REJdu1                2.827201                3.898633   
0017A9myIXhXYB4gBsPRsA                2.235485                3.037384   
001reycpvPkUjmpxTJv6eV                2.916704                2.927719   
002H0J6N0wKjjRtFQhKDzZ                4.913558                2.686284   

spotify_id              003787JNriexFRPKRwyMTn  003FCWl5ci9MxCaiTbErSr  \
spotify_id                                                               
000P83HDtOHcNVFZy7Q2Yu                2.155379                3.570523   
000paUq6xoYcvCz2REJdu1                3.234504                4.414538   
0017A9myIXhXYB4gBsPRsA                2.455009                3.526743   
001reycpvPkUjmpxTJv6eV                3.064924                3.243758   
002H0J6N0wKjjRtFQhKDzZ                2.265792                3.246359   

spotify_id              004223m8Zb9ZHcxHFxlpQc  006XjVYo0HbGrUfvL4FodP  ...  \
spotify_id                                                              ...   
000P83HDtOHcNVFZy7Q2Yu                3.197774                2.039307  ...   
000paUq6xoYcvCz2REJdu1                2.987895                3.091287  ...   
0017A9myIXhXYB4gBsPRsA                2.639685                2.276847  ...   
001reycpvPkUjmpxTJv6eV                2.612997                3.233974  ...   
002H0J6N0wKjjRtFQhKDzZ                2.864889                1.883516  ...   

spotify_id              7zn1gpif2DHi9t9isA5Boe  7znOzcWy38z63lIcKOQBd4  \
spotify_id                                                               
000P83HDtOHcNVFZy7Q2Yu                3.181161                2.756306   
000paUq6xoYcvCz2REJdu1                4.115933                3.525990   
0017A9myIXhXYB4gBsPRsA                3.557461                3.144690   
001reycpvPkUjmpxTJv6eV                2.963520                2.347248   
002H0J6N0wKjjRtFQhKDzZ                3.011848                2.388455   

spotify_id              7zndxriU3fd1TkV9BMdv5Y  7zraaKkjSygFzRAObStH8C  \
spotify_id                                                               
000P83HDtOHcNVFZy7Q2Yu                3.610660                3.685095   
000paUq6xoYcvCz2REJdu1                6.358596                4.395858   
0017A9myIXhXYB4gBsPRsA                4.401928                3.829674   
001reycpvPkUjmpxTJv6eV                3.288713                3.067704   
002H0J6N0wKjjRtFQhKDzZ                3.264198                3.273877   

spotify_id              7zrlCQpCrfAxWEop9J8OzK  7zsH8mPq4nPBqqqDJpjgJe  \
spotify_id                                                               
000P83HDtOHcNVFZy7Q2Yu                2.730652                2.445887   
000paUq6xoYcvCz2REJdu1                3.642021                3.197646   
0017A9myIXhXYB4gBsPRsA              

In [45]:
df.sort_values('popularity', ascending=False).head(10)

name                   artists  \
239560                Back In Black                     AC/DC   
126862                Back In Black                     AC/DC   
237518  You Shook Me All Night Long                     AC/DC   
126560  You Shook Me All Night Long                     AC/DC   
236291           Could You Be Loved  Bob Marley & The Wailers   
196916                       9 to 5              Dolly Parton   
65758                        9 to 5              Dolly Parton   
126562                Ace of Spades                 Motörhead   
151771                Ace of Spades                 Motörhead   
287176      The Winner Takes It All                      ABBA   

                                   album release_date release_date_precision  \
239560                     Back In Black   1980-07-25                    day   
126862                     Back In Black   1980-07-25                    day   
237518                     Back In Black   1980-07-25                    day   
126560                     Back In Black   1980-07-25                    day   
236291                          Uprising         1980                   year   
196916               9 To 5 And Odd Jobs         1980                   year   
65758                9 To 5 And Odd Jobs         1980                   year   
126562  Ace of Spades (Expanded Edition)   1980-11-08                    day   
151771  Ace of Spades (Expanded Edition)   1980-11-08                    day   
287176                     Super Trouper         1980                   year   

        chart_power              spotify_id  \
239560          NaN  08mG3Y1vljYA6bvDt4Wqkj   
126862          NaN  08mG3Y1vljYA6bvDt4Wqkj   
237518        901.0  2SiXAy7TuUkycRVbbWDEpo   
126560        901.0  2SiXAy7TuUkycRVbbWDEpo   
236291       1823.0  5O4erNlJ74PIF6kGol1ZrC   
196916        548.0  4w3tQBXhn5345eUXDGBWZG   
65758         548.0  4w3tQBXhn5345eUXDGBWZG   
126562          NaN  6EPRKhUOdiFSQwGBRBbvsZ   
151771          NaN  6EPRKhUOdiFSQwGBRBbvsZ   
287176       1761.0  3oEkrIfXfSh9zGnE7eBzSV   

                                         uri  popularity  danceability  ...  \
239560  spotify:track:08mG3Y1vljYA6bvDt4Wqkj          82         0.310  ...   
126862  spotify:track:08mG3Y1vljYA6bvDt4Wqkj          82         0.310  ...   
237518  spotify:track:2SiXAy7TuUkycRVbbWDEpo          81         0.532  ...   
126560  spotify:track:2SiXAy7TuUkycRVbbWDEpo          80         0.532  ...   
236291  spotify:track:5O4erNlJ74PIF6kGol1ZrC          79         0.916  ...   
196916  spotify:track:4w3tQBXhn5345eUXDGBWZG          77         0.554  ...   
65758   spotify:track:4w3tQBXhn5345eUXDGBWZG          76         0.554  ...   
126562  spotify:track:6EPRKhUOdiFSQwGBRBbvsZ          74         0.342  ...   
151771  spotify:track:6EPRKhUOdiFSQwGBRBbvsZ          74         0.342  ...   
287176  spotify:track:3oEkrIfXfSh9zGnE7eBzSV          74         0.454  ...   

        speechiness  acousticness  instrumentalness  liveness  valence  \
239560       0.0470       0.01100          0.009650    0.0828    0.763   
126862       0.0470       0.01100          0.009650    0.0828    0.763   
237518       0.0574       0.00287          0.000513    0.3900    0.755   
126560       0.0574       0.00287          0.000513    0.3900    0.755   
236291       0.1000       0.36000          0.000160    0.0958    0.760   
196916       0.0457       0.41600          0.000002    0.6310    0.813   
65758        0.0457       0.41600          0.000002    0.6310    0.813   
126562       0.0758       0.00004          0.000121    0.0823    0.547   
151771       0.0758       0.00004          0.000121    0.0823    0.547   
287176       0.0371       0.57400          0.000766    0.0859    0.520   

          tempo  duration_ms  time_signature          isrc     genres  
239560  188.386       255493               4  AUAP08000046       rock  
126862  188.386       255493               4  AUAP08000046  hard-rock  
237518  127.361

In [46]:
def recommend_tracks(track: str, df_rec: pd.DataFrame):
    '''
    Recommends tracks that are similar to the provided track.

    Parameter
    ---------
    track: str
        Provided track

    df: pd.DataFrame
        DataFrame used for the Recommendation
    
    '''
    ids = df[df.name == track]['spotify_id'].values
    for id in ids:
        recommender_df = df_rec[[ids[0]]]
        recommender_df.reset_index(inplace=True)
        recommender_df = recommender_df.rename(columns={id: 'Score'})
        recommender_df = recommender_df.merge(df_total, how='left', on='spotify_id')
        display(recommender_df.sort_values(by='Score', ascending=False).head(10))


In [47]:
recommend_tracks('Back In Black', cosine_similarity_df)

spotify_id     Score             genres  \
722    08mG3Y1vljYA6bvDt4Wqkj  1.000000  [hard-rock, rock]   
25422  5AD7oNm0kvWIXxW3vy0z2a  0.996716  [hard-rock, rock]   
26519  5O4rVUjsG66l2nArhYW57m  0.996298  [hard-rock, rock]   
6199   1En7X8Ama1U8MosJbDndnB  0.992350  [hard-rock, rock]   
18529  3m3hmciMkd4y3XGC1mjTHf  0.991354  [hard-rock, rock]   
36341  7MxmJmW08fXXIx0WV1UeKz  0.990243  [hard-rock, rock]   
25670  5DVXOyNMl4MunWZ2BDvgA5  0.989644  [hard-rock, rock]   
25215  57hFsLrkALFFIxncLAY7Tg  0.989644  [hard-rock, rock]   
16911  3R306PKKOPAUjSSMkDHOpY  0.988208  [hard-rock, rock]   
21489  4NyZk1OfjwkGiIQxsD10ii  0.987438  [hard-rock, rock]   

                                                    name  \
722                       [Back In Black, Back In Black]   
25422                     [Homemade Love, Homemade Love]   
26519                           [Relentless, Relentless]   
6199                        [Too Far Gone, Too Far Gone]   
18529                 [Rolê da Estrada, Rolê da Estrada]   
36341                 [Someone Tonight, Someone Tonight]   
25670                     [Hello America, Hello America]   
25215                     [Hello America, Hello America]   
16911  [Rock n' Roll Is Alive and Well, Rock n' Roll ...   
21489                                 [Anytime, Anytime]   

                                        artists  \
722                              [AC/DC, AC/DC]   
25422                        [Journey, Journey]   
26519                          [Kansas, Kansas]   
6199                     [The Babys, The Babys]   
18529  [Patrulha do Espaço, Patrulha do Espaço]   
36341          [REO Speedwagon, REO Speedwagon]   
25670                [Def Leppard, Def Leppard]   
25215                [Def Leppard, Def Leppard]   
16911                    [The Babys, The Babys]   
21489                    [The Babys, The Babys]   

                                                   album  \
722                       [Back In Black, Back In Black]   
25422                             [Departure, Departure]   
26519                     [Audio-Visions, Audio-Visions]   
6199                          [On the Edge, On the Edge]   
18529           [Patrulha do Espaço, Patrulha do Espaço]   
36341  [Hi Infidelity (30th Anniversary Edition), Hi ...   
25670  [On Through The Night (Remastered), On Through...   
25215       [On Through The Night, On Through The Night]   
16911                         [On the Edge, On the Edge]   
21489                         [Union Jacks, Union Jacks]   

                   release_date release_date_precision  \
722    [1980-07-25, 1980-07-25]             [day, day]   
25422              [1980, 1980]           [year, year]   
26519  [1980-09-01, 1980-09-01]             [day, day]   
6199   [1980-10-01, 1980-10-01]             [day, day]   
18529  [1980-10-10, 1980-10-10]             [day, day]   
36341              [1980, 1980]           [year, year]   
25670  [1980-03-14, 1980-03-14]             [day, day]   
25215  [1980-03-14, 1980-03-14]             [day, day]   
16911  [1980-10-01, 1980-10-01]             [day, day]   
21489  [1980-01-01, 1980-01-01]             [day, day]   

                                                     uri  \
722    [spotify:track:08mG3Y1vljYA6bvDt4Wqkj, spotify...   
25422  [spotify:track:5AD7oNm0kvWIXxW3vy0z2a, spotify...   
26519  [spotify:track:5O4rVUjsG66l2nArhYW57m, spotify...   
6199   [spotify:track:1En7X8Ama1U8MosJbDndnB, spotify...   
18529  [spotify:track:3m3hmciMkd4y3XGC1mjTHf, spotify...   
36341  [spotify:track:7MxmJmW08fXXIx0WV1UeKz, spotify...   
25670  [spotify:track:5DVXOyNMl4MunWZ2BDvgA5, spotify...   
25215  [spotify:track:57hFsLrkALFFIxncLAY7Tg, spotify...   
16911  [spotify:track:3R306PKKOPAUjSSMkDHOpY, spotify...   
21489  [spotify:track:4NyZk1OfjwkGiIQxsD10ii, spotify...   

                               isrc  ...  loudness  mode  speechiness  \
722    [AUAP08000046, AUAP08000046]  ...    -5.678   1.0       0.0470   
25422 

spotify_id     Score             genres  \
722    08mG3Y1vljYA6bvDt4Wqkj  1.000000  [hard-rock, rock]   
25422  5AD7oNm0kvWIXxW3vy0z2a  0.996716  [hard-rock, rock]   
26519  5O4rVUjsG66l2nArhYW57m  0.996298  [hard-rock, rock]   
6199   1En7X8Ama1U8MosJbDndnB  0.992350  [hard-rock, rock]   
18529  3m3hmciMkd4y3XGC1mjTHf  0.991354  [hard-rock, rock]   
36341  7MxmJmW08fXXIx0WV1UeKz  0.990243  [hard-rock, rock]   
25670  5DVXOyNMl4MunWZ2BDvgA5  0.989644  [hard-rock, rock]   
25215  57hFsLrkALFFIxncLAY7Tg  0.989644  [hard-rock, rock]   
16911  3R306PKKOPAUjSSMkDHOpY  0.988208  [hard-rock, rock]   
21489  4NyZk1OfjwkGiIQxsD10ii  0.987438  [hard-rock, rock]   

                                                    name  \
722                       [Back In Black, Back In Black]   
25422                     [Homemade Love, Homemade Love]   
26519                           [Relentless, Relentless]   
6199                        [Too Far Gone, Too Far Gone]   
18529                 [Rolê da Estrada, Rolê da Estrada]   
36341                 [Someone Tonight, Someone Tonight]   
25670                     [Hello America, Hello America]   
25215                     [Hello America, Hello America]   
16911  [Rock n' Roll Is Alive and Well, Rock n' Roll ...   
21489                                 [Anytime, Anytime]   

                                        artists  \
722                              [AC/DC, AC/DC]   
25422                        [Journey, Journey]   
26519                          [Kansas, Kansas]   
6199                     [The Babys, The Babys]   
18529  [Patrulha do Espaço, Patrulha do Espaço]   
36341          [REO Speedwagon, REO Speedwagon]   
25670                [Def Leppard, Def Leppard]   
25215                [Def Leppard, Def Leppard]   
16911                    [The Babys, The Babys]   
21489                    [The Babys, The Babys]   

                                                   album  \
722                       [Back In Black, Back In Black]   
25422                             [Departure, Departure]   
26519                     [Audio-Visions, Audio-Visions]   
6199                          [On the Edge, On the Edge]   
18529           [Patrulha do Espaço, Patrulha do Espaço]   
36341  [Hi Infidelity (30th Anniversary Edition), Hi ...   
25670  [On Through The Night (Remastered), On Through...   
25215       [On Through The Night, On Through The Night]   
16911                         [On the Edge, On the Edge]   
21489                         [Union Jacks, Union Jacks]   

                   release_date release_date_precision  \
722    [1980-07-25, 1980-07-25]             [day, day]   
25422              [1980, 1980]           [year, year]   
26519  [1980-09-01, 1980-09-01]             [day, day]   
6199   [1980-10-01, 1980-10-01]             [day, day]   
18529  [1980-10-10, 1980-10-10]             [day, day]   
36341              [1980, 1980]           [year, year]   
25670  [1980-03-14, 1980-03-14]             [day, day]   
25215  [1980-03-14, 1980-03-14]             [day, day]   
16911  [1980-10-01, 1980-10-01]             [day, day]   
21489  [1980-01-01, 1980-01-01]             [day, day]   

                                                     uri  \
722    [spotify:track:08mG3Y1vljYA6bvDt4Wqkj, spotify...   
25422  [spotify:track:5AD7oNm0kvWIXxW3vy0z2a, spotify...   
26519  [spotify:track:5O4rVUjsG66l2nArhYW57m, spotify...   
6199   [spotify:track:1En7X8Ama1U8MosJbDndnB, spotify...   
18529  [spotify:track:3m3hmciMkd4y3XGC1mjTHf, spotify...   
36341  [spotify:track:7MxmJmW08fXXIx0WV1UeKz, spotify...   
25670  [spotify:track:5DVXOyNMl4MunWZ2BDvgA5, spotify...   
25215  [spotify:track:57hFsLrkALFFIxncLAY7Tg, spotify...   
16911  [spotify:track:3R306PKKOPAUjSSMkDHOpY, spotify...   
21489  [spotify:track:4NyZk1OfjwkGiIQxsD10ii, spotify...   

                               isrc  ...  loudness  mode  speechiness  \
722    [AUAP08000046, AUAP08000046]  ...    -5.678   1.0       0.0470   
25422 

In [48]:
recommend_tracks('Back In Black', linear_kernel_df)

spotify_id     Score  \
2112   0PLzoe4izqluPTVsF9K4mb  7.155934   
4650   0vEEhp1BFWt1iCKsojoebd  7.150820   
6199   1En7X8Ama1U8MosJbDndnB  7.114456   
1578   0Ix7jl28ovNUGclUmkwHnx  7.092672   
28209  5joQhLd0l0jc1S0U9TkmsG  7.073923   
29747  63YhDDyYQJo847NqN7mry3  7.066416   
17610  3Zw7u9z5sfbasLy9zkubzn  7.045456   
30396  6BUuoR1kG1rRag718ibuOn  7.043061   
12883  2cXADTZcYNhvtfZuK77Xt0  7.040993   
1883   0MLxcHeTU7mJWAKbbwOPT5  7.034511   

                                                  genres  \
2112        [hard-rock, hardcore, punk, punk-rock, rock]   
4650        [hard-rock, hardcore, punk, punk-rock, rock]   
6199                                   [hard-rock, rock]   
1578                                   [hard-rock, rock]   
28209  [alt-rock, alternative, blues, hard-rock, hard...   
29747  [alt-rock, alternative, blues, grunge, hard-ro...   
17610           [blues, british, hard-rock, metal, rock]   
30396                   [country, folk, hard-rock, rock]   
12883                    [blues, hard-rock, metal, rock]   
1883   [alt-rock, alternative, hard-rock, hardcore, p...   

                                                    name  \
2112   [Ennen kolmatta maailmansotaa, Ennen kolmatta ...   
4650                 [Ska 4, Ska 4, Ska 4, Ska 4, Ska 4]   
6199                        [Too Far Gone, Too Far Gone]   
1578                                  [My Life, My Life]   
28209  [Soul Kitchen, Soul Kitchen, Soul Kitchen, Sou...   
29747  [Cover Band - Demo, Cover Band - Demo, Cover B...   
17610  [Carry Your Load, Carry Your Load, Carry Your ...   
30396  [Betty Lou's Gettin' Out Tonight, Betty Lou's ...   
12883  [Hungry Boys, Hungry Boys, Hungry Boys, Hungry...   
1883   [Let's Go Die, Let's Go Die, Let's Go Die, Let...   

                                                 artists  \
2112   [Maukka Perusjätkä,Ralf Örn Pop, Maukka Perusj...   
4650   [Maukka Perusjätkä,Ralf Örn Pop, Maukka Perusj...   
6199                              [The Babys, The Babys]   
1578                          [Peter Criss, Peter Criss]   
28209                        [X, X, X, X, X, X, X, X, X]   
29747  [Redd Kross, Redd Kross, Redd Kross, Redd Kros...   
17610  [Whitesnake, Whitesnake, Whitesnake, Whitesnak...   
30396       [Bob Seger, Bob Seger, Bob Seger, Bob Seger]   
12883  [Blue Öyster Cult, Blue Öyster Cult, Blue Öyst...   
1883   [Hüsker Dü, Hüsker Dü, Hüsker Dü, Hüsker Dü, H...   

                                                   album  \
2112   [Ennen kolmatta maailmansotaa, Ennen kolmatta ...   
4650   [Ennen kolmatta maailmansotaa, Ennen kolmatta ...   
6199                          [On the Edge, On the Edge]   
1578                    [Out Of Control, Out Of Control]   
28209  [Los Angeles, Los Angeles, Los Angeles, Los An...   
29747  [Red Cross, Red Cross, Red Cross, Red Cross, R...   
17610  [Ready An' Willing, Ready An' Willing, Ready A...   
30396  [Against The Wind, Against The Wind, Against T...   
12883  [Cultosaurus Erectus, Cultosaurus Erectus, Cul...   
1883   [Statues / Amusement, Statues / Amusement, Sta...   

                                            release_date  \
2112   [1980-12-12, 1980-12-12, 1980-12-12, 1980-12-1...   
4650   [1980-12-12, 1980-12-12, 1980-12-12, 1980-12-1...   
6199                            [1980-10-01, 1980-10-01]   
1578                                        [1980, 1980]   
28209  [1980-04-26, 1980-04-26, 1980-04-26, 1980-04-2...   
29747  [1980-01-01, 1980-01-01, 1980-01-01, 1980-01-0...   
17610  [1980-01-01, 1980-01-01, 1980-01-01, 1980-01-0...   
30396   [1980-02-25, 1980-02-25, 1980-02-25, 1980-02-25]   
12883                           [1980, 1980, 1980, 1980]   
1883   [1980-08-01, 1980-08-01, 1980-08-01, 1980-08-0...   

                                  release_date_precision  \
2112                           [day, day, day, day, day]   
4650                           [day, day, day, day, day]   
6199                                          [day, day]   


spotify_id     Score  \
2112   0PLzoe4izqluPTVsF9K4mb  7.155934   
4650   0vEEhp1BFWt1iCKsojoebd  7.150820   
6199   1En7X8Ama1U8MosJbDndnB  7.114456   
1578   0Ix7jl28ovNUGclUmkwHnx  7.092672   
28209  5joQhLd0l0jc1S0U9TkmsG  7.073923   
29747  63YhDDyYQJo847NqN7mry3  7.066416   
17610  3Zw7u9z5sfbasLy9zkubzn  7.045456   
30396  6BUuoR1kG1rRag718ibuOn  7.043061   
12883  2cXADTZcYNhvtfZuK77Xt0  7.040993   
1883   0MLxcHeTU7mJWAKbbwOPT5  7.034511   

                                                  genres  \
2112        [hard-rock, hardcore, punk, punk-rock, rock]   
4650        [hard-rock, hardcore, punk, punk-rock, rock]   
6199                                   [hard-rock, rock]   
1578                                   [hard-rock, rock]   
28209  [alt-rock, alternative, blues, hard-rock, hard...   
29747  [alt-rock, alternative, blues, grunge, hard-ro...   
17610           [blues, british, hard-rock, metal, rock]   
30396                   [country, folk, hard-rock, rock]   
12883                    [blues, hard-rock, metal, rock]   
1883   [alt-rock, alternative, hard-rock, hardcore, p...   

                                                    name  \
2112   [Ennen kolmatta maailmansotaa, Ennen kolmatta ...   
4650                 [Ska 4, Ska 4, Ska 4, Ska 4, Ska 4]   
6199                        [Too Far Gone, Too Far Gone]   
1578                                  [My Life, My Life]   
28209  [Soul Kitchen, Soul Kitchen, Soul Kitchen, Sou...   
29747  [Cover Band - Demo, Cover Band - Demo, Cover B...   
17610  [Carry Your Load, Carry Your Load, Carry Your ...   
30396  [Betty Lou's Gettin' Out Tonight, Betty Lou's ...   
12883  [Hungry Boys, Hungry Boys, Hungry Boys, Hungry...   
1883   [Let's Go Die, Let's Go Die, Let's Go Die, Let...   

                                                 artists  \
2112   [Maukka Perusjätkä,Ralf Örn Pop, Maukka Perusj...   
4650   [Maukka Perusjätkä,Ralf Örn Pop, Maukka Perusj...   
6199                              [The Babys, The Babys]   
1578                          [Peter Criss, Peter Criss]   
28209                        [X, X, X, X, X, X, X, X, X]   
29747  [Redd Kross, Redd Kross, Redd Kross, Redd Kros...   
17610  [Whitesnake, Whitesnake, Whitesnake, Whitesnak...   
30396       [Bob Seger, Bob Seger, Bob Seger, Bob Seger]   
12883  [Blue Öyster Cult, Blue Öyster Cult, Blue Öyst...   
1883   [Hüsker Dü, Hüsker Dü, Hüsker Dü, Hüsker Dü, H...   

                                                   album  \
2112   [Ennen kolmatta maailmansotaa, Ennen kolmatta ...   
4650   [Ennen kolmatta maailmansotaa, Ennen kolmatta ...   
6199                          [On the Edge, On the Edge]   
1578                    [Out Of Control, Out Of Control]   
28209  [Los Angeles, Los Angeles, Los Angeles, Los An...   
29747  [Red Cross, Red Cross, Red Cross, Red Cross, R...   
17610  [Ready An' Willing, Ready An' Willing, Ready A...   
30396  [Against The Wind, Against The Wind, Against T...   
12883  [Cultosaurus Erectus, Cultosaurus Erectus, Cul...   
1883   [Statues / Amusement, Statues / Amusement, Sta...   

                                            release_date  \
2112   [1980-12-12, 1980-12-12, 1980-12-12, 1980-12-1...   
4650   [1980-12-12, 1980-12-12, 1980-12-12, 1980-12-1...   
6199                            [1980-10-01, 1980-10-01]   
1578                                        [1980, 1980]   
28209  [1980-04-26, 1980-04-26, 1980-04-26, 1980-04-2...   
29747  [1980-01-01, 1980-01-01, 1980-01-01, 1980-01-0...   
17610  [1980-01-01, 1980-01-01, 1980-01-01, 1980-01-0...   
30396   [1980-02-25, 1980-02-25, 1980-02-25, 1980-02-25]   
12883                           [1980, 1980, 1980, 1980]   
1883   [1980-08-01, 1980-08-01, 1980-08-01, 1980-08-0...   

                                  release_date_precision  \
2112                           [day, day, day, day, day]   
4650                           [day, day, day, day, day]   
6199                                          [day, day]   
